hudi-metastore/src/main/java/org/apache/hudi/metastore/client/HoodieMetastoreClientImp.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.metastore.client;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hudi.common.config.HoodieMetastoreConfig;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.metastore.HoodieMetastoreServer;
+import org.apache.hudi.metastore.thrift.Table;
+import org.apache.hudi.metastore.thrift.ThriftHoodieMetastore;
+import org.apache.hudi.metastore.util.EntryConvertor;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.transport.TSocket;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class HoodieMetastoreClientImp implements HoodieMetastoreClient, Serializable {
+
+  private static final Logger LOG =  LogManager.getLogger(HoodieMetastoreClientImp.class);
+  private final HoodieMetastoreConfig config;
+  private final int retryLimit;
+  private final int retryDelaySeconds;
+  private boolean isConnected;
+  private boolean isLocal;
+  private ThriftHoodieMetastore.Iface client;
+  private TTransport transport;
+
+  public HoodieMetastoreClientImp(HoodieMetastoreConfig config) {
+    this.config = config;
+    this.retryLimit = config.getConnectionRetryLimit();
+    this.retryDelaySeconds = config.getConnectionRetryDelay();
+    String uri = config.getMetastoreUris();
+    if (isLocalEmbeddedMetastore(uri)) {
+      this.client = HoodieMetastoreServer.getEmbeddedMetastore();
+      this.isConnected = true;
+      this.isLocal = true;
+    } else {
+      open();
+    }
+  }
+
+  private void open() {
+    String uri = config.getMetastoreUris();
+    TTransportException exception = null;
+    for (int i = 0; i < retryLimit; i++) {
```

Comment:
for (int 1= 0; !isConnected && 1< retryLimit; 1++)

Reply:
nice catch